zkvm/src/encoding.rs

```diff
@@ -44,26 +43,45 @@ impl<'a> Subslice<'a> {
         self.start..self.end
     }
 
-    /// Copies the contents of the subslice into a `Vec<u8>`.
-    pub fn to_vec(&self) -> Vec<u8> {
-        let mut buf = Vec::with_capacity(self.len());
-        buf.extend_from_slice(&self);
-        buf
+    pub fn slice<F, T>(&mut self, slice_fn: F) -> Result<(T, &[u8]), VMError>
+    where
+        F: FnOnce(&mut SliceReader) -> Result<T, VMError>,
+    {
+        let start = self.start;
+        let result = slice_fn(self)?;
+        let end = self.start;
+        Ok((result, &self.whole[start..end]))
     }
 
-    /// Returns a Subslice of the first `prefix_size` of bytes and advances
+    pub fn parse<F, T>(data: &[u8], parse_fn: F) -> Result<T, VMError>
+    where
+        F: FnOnce(SliceReader) -> Result<T, VMError>,
+    {
+        let reader = SliceReader::new(data);
+        let result = parse_fn(reader)?;
+        if reader.len() != 0 {
+            return Err(VMError::FormatError);
+        }
+        Ok(result)
+    }
+
+    pub fn skip_trailing_bytes(&mut self) -> usize {
+        unimplemented!()
+    }
+
+    /// Returns a slice of the first `prefix_size` of bytes and advances
     /// the internal offset.
-    pub fn read_bytes(&mut self, prefix_size: usize) -> Result<Subslice, VMError> {
+    pub fn read_bytes(&mut self, prefix_size: usize) -> Result<&[u8], VMError> {
         if prefix_size > self.len() {
             return Err(VMError::FormatError);
         }
-        let prefix = Subslice {
+        let prefix = SliceReader {
```

Comment:
We probably don't even need to construct this instance if we return a raw slice. And we can kill the `range()` method entirely too, i think.

How about:

```rust
pub fn read_bytes(&mut self, prefix_size: usize) -> Result<&[u8], VMError> {
    if prefix_size > self.len() {
        return Err(VMError::FormatError);
    }
    let prefix = self.whole[self.start..(self.start + prefix_size)];
    self.start += prefix_size;
    Ok(prefix)
}
```